### PR TITLE
[sparse] add sparsify rule for lax.cond

### PR DESCRIPTION
**Pull request by jakevdp**

### CODE REVIEW

**Review by froystig — COMMENTED** [review]
A requirement of our `cond` primitive is that all branches have the same signature. I think the current rule produces branches whose input types are the same, but does it also guarantee a uniform output type? What if the output array is sparse in one branch but not the other?

I'm thinking of a test case along the following lines:
```python
def f(x, y):
  return lax.cond(pred, lambda z: z[0], lambda z: z[1], (x, y))
```
sparsified only on the first argument.

File: jax/experimental/sparse/transform.py [hunk 1]
```diff
@@ -423,6 +423,12 @@ def _xla_call_sparse(spenv, *argspecs, call_jaxpr, donated_invars, **params):
 
 sparse_rules[xla.xla_call_p] = _xla_call_sparse
 
+
+def _map_params(argspecs, params):
```
[the remark this answers is not in the document]
Comment (jakevdp) [reply jax/experimental/sparse/transform.py]: Done

**Comment by jakevdp** [comment]
Regarding mismatched cond outputs: I thought about that, and I believe raising an exception in this case is the correct result. I believe the existing error checking within `cond_p.bind` will raise appropriate errors if the sparsified branch jaxprs don't match.

**Comment by froystig** [comment]
Do we have the same error check within `cond_p.bind` in mind? This one that ensures [uniform output types](https://github.com/google/jax/blob/856c88fe44a6b658e6db9c8956b7ca64012d1d09/jax/_src/lax/control_flow.py#L1101-L1104) is [conditional](https://github.com/google/jax/blob/856c88fe44a6b658e6db9c8956b7ca64012d1d09/jax/_src/lax/control_flow.py#L1121-L1123), enabled by a config flag, so I'm not sure that we can rely on it. Its error message might also confuse here, since in the input program (the one possibly written by the user) the cond branches do all match.

**Comment by jakevdp** [comment]
OK, added an explicit check & a test.

**Review by froystig — APPROVED** [review]
Nice – it looks good!

**Pull request merged**